### ISSUE TITLE
Ошибка при наследовании модуля плагина работающего на орм и с указанием кастомных имен таблиц через конфиг

### ISSUE THREAD

**Issue by psnet** [livestreet/livestreet-framework#150]
По-умолчанию таблицы орм для плагинов в лс 2.0 будут именоваться не так как в 1.0.3, а с именем префикса плагина, то этот баг для сложных проектов был бы *очень веселой штукой*.

Условия задачи
---
Наследовать орм модуль плагина, который дает свои имена таблицам плагина через конфиг. Чтобы было более понятно, то пример будет такой: есть плагин `A` с орм модулем и сущностью:
```php
PluginA_ModuleA
PluginA_ModuleA_EntityA
```
и есть плагин `B`, который наследует **только** модуль плагина `A`:
```php
PluginB_ModuleB -> PluginA_ModuleA
```
и для плагина A в конфиге указаны кастомные имена таблиц для сущностей.

Поехали
---

В орм в методах таких как `GetByFilter`, `GetItemsByFilter` и других есть конструкции:
```php
if (is_null($sEntityFull)) {
  $sEntityFull=Engine::GetPluginPrefix($this).'Module'.Engine::GetModuleName($this).'_Entity'.Engine::GetModuleName(get_class($this));
} elseif (!substr_count($sEntityFull,'_')) {
  $sEntityFull=Engine::GetPluginPrefix($this).'Module'.Engine::GetModuleName($this).'_Entity'.$sEntityFull;
}
```
Нетрудно догадаться что при наследовании модуля плагина <strong>А</strong> переменная <strong>$sEntityFull</strong> получит префикс плагина <strong>В</strong>, который наследует исходный и в итоге будет получена сущность, которая реально не существует вообще - <strong>PluginB_ModuleB_EntityA</strong>. Далее эта переменная передается в маппер, где происходит получение этой не существующей сущности:
```php
$oEntitySample=Engine::GetEntity($sEntityFull);
```
И тут вполне ожидаемо было бы ждать ексепшн, т.к. сущности такой нет, это ошибка, но тут работает самый давний, самый натуральный что ни есть хак, на который я всегда косо смотрел:
```php
/**
 * If Plugin Entity doesn't exist, search among it's Module delegates
 */
if(isset($sPlugin) && !self::GetClassPath($sClass)) {
  $aModulesChain = Engine::GetInstance()->Plugin_GetDelegationChain('module','Plugin'.$sPlugin.'_Module'.$sModule);
  foreach($aModulesChain as $sModuleName) {
    $sClassTest=$sModuleName.'_Entity'.$sEntity;
    if(self::GetClassPath($sClassTest)) {
      $sClass=$sClassTest;
      break;
    }
  }
  if(!self::GetClassPath($sClass)) {
    $sClass='Module'.$sModule.'_Entity'.$sEntity;
  }
}
```
И ядро находит таки сущность среди сущностей делегируемых модулей плагина. Уже на данном этапе это проблема в архитектуре, которая должна быть решена, это неверное поведение.

Далее вызов обратно возвращается в орм маппер, где идет получение таблицы, но не с полученной корректной сущности от ядра, а снова от её (ошибочного) имени:
```php
$sTableName = self::GetTableName($sEntityFull);
```
естественно метод `GetTableName` разбирает имя на составляющие и не найдет таблицу исходного плагина <strong>A</strong> для орм модуля, т.к. будет искать с именем плагина <strong>B</strong>, который наследует оригинальный и работа завершится ошибкой не нахождения таблицы.

Варианты решения
---

- Заменить в орм маппере движка вызовы
```php
$sTableName = self::GetTableName($sEntityFull);
```
на
```php
$sTableName = self::GetTableName($oEntitySample);
```
это как минимум должно быть сделано в движке для новой версии лс. Примечательно что в некоторых методах именно так и сделано, например, в методе получения полей таблицы `ShowColumnsFrom`, получения примари ключа `ShowPrimaryIndexFrom` или просто добавления/обновления сущностей, но там изначально передается в методы сама сущность.

- Сделать правку выше + исправить все места формирования сущности таким образом, чтобы избавится от вышеуказанного хака в движке и вынести формирование сущности:
```php
if (is_null($sEntityFull)) {
  $sEntityFull=Engine::GetPluginPrefix($this).'Module'.Engine::GetModuleName($this).'_Entity'.Engine::GetModuleName(get_class($this));
} elseif (!substr_count($sEntityFull,'_')) {
  $sEntityFull=Engine::GetPluginPrefix($this).'Module'.Engine::GetModuleName($this).'_Entity'.$sEntityFull;
}
```
в методах орм в отдельный метод, который и получит корректную сущность, даже если он будет содержать тот же код хака из ядра, но это будет значительно лучше т.к. дальше по коду в методах орм (например, все тот же `GetItemsByFilter`) идет сессионный кеш:
```php
$sEntityFullRoot=$this->Plugin_GetRootDelegater('entity',$sEntityFull);
$sCacheKey=$sEntityFullRoot.'_items_by_filter_'.serialize($aFilter);
$aCacheTags=array($sEntityFullRoot.'_save',$sEntityFullRoot.'_delete');
```
и **он строится на неверном имени сущности и ключ и теги кеша получаются тоже не корректные** (`GetRootDelegater` вернет тоже имя т.к. нету такой сущности и нет, соответственно, исходной).

- А разработчикам придется писать в своих плагинах обходной временной хак с маппингом имени таблицы для каждой сущности оригинального плагина (<strong>А</strong>) на имя плагина-наследника (<strong>В</strong>):
```php
$config['$root$']['db']['table']['ПлагинB_МодульB_СущностьА'] = '___db.table.ПлагинА_МодульА_СущностьА___';
```
<a href="http://livestreetguide.com/dev/oshibka-ne-nahozhdeniya-tablicy-pri-nasledovanii-modulya-plagina-rabotayuschego-na-orm-i-s-ukazaniem-kastomnyh-imen-tablic-cherez-konfig.html">Оригинал статьи</a>

**Comment by psnet** [comment]
Хочу заметить, что данный баг критичен и потребует выпуска фикса для первой версии лс т.к. теги кеша орм сущностей (`сущность_save`, `сущность_delete`) формируют с ошибкой и поэтому при включенном кешировании кеш плагина частично не будет сброшен и это приведет к самым неожиданным ситуациям ("зависание" некоторых объектов и др.)

**Comment by mzhelskiy** [comment]
По тексту так и не понял - это актуально для 2.0 или нет?

**Comment by psnet** [comment]
Нет, в 2.0 уже исправлено в т.ч. и вынесен код формирования полного имени сущности в отдельный метод `_NormalizeEntityRootName` который корректно получает сущность благодаря поэтапному получению корневого класса модуля и сущности:
```php
        /**
         * Получаем корневой модуль
         */
        $sModuleRoot = $this->Plugin_GetRootDelegater('module', $sPluginPrefix . 'Module' . $sModuleName);
        /**
         * Возвращаем корневую сущность
         */
        return $this->Plugin_GetRootDelegater('entity', $sModuleRoot . '_Entity' . $sEntityName);
```
а не всей сущности сразу как в 1*:
```php
$sEntityFullRoot=$this->Plugin_GetRootDelegater('entity',$sEntityFull);
```

**Comment by psnet** [comment]
Но в таком случае для 2.0 также нужно избавиться от хака в `Engine::GetEntity`:
```php
/**
 * If Plugin Entity doesn't exist, search among it's Module delegates
 */
if(isset($sPlugin) && !self::GetClassPath($sClass)) {
  $aModulesChain = Engine::GetInstance()->Plugin_GetDelegationChain('module','Plugin'.$sPlugin.'_Module'.$sModule);
  foreach($aModulesChain as $sModuleName) {
    $sClassTest=$sModuleName.'_Entity'.$sEntity;
    if(self::GetClassPath($sClassTest)) {
      $sClass=$sClassTest;
      break;
    }
  }
  if(!self::GetClassPath($sClass)) {
    $sClass='Module'.$sModule.'_Entity'.$sEntity;
  }
}
```
